catalyst-toolbox-cli/src/cli/recovery/tally/mockchain.rs

```diff
@@ -0,0 +1,86 @@
+use chain_impl_mockchain::block::HeaderId;
+use chain_impl_mockchain::fee::LinearFee;
+use chain_impl_mockchain::fragment::Fragment;
+use chain_impl_mockchain::ledger::{Error, Ledger, LedgerParameters};
+use chain_impl_mockchain::rewards::TaxType;
+use jormungandr_lib::interfaces::PersistentFragmentLog;
+
+use std::collections::vec_deque::VecDeque;
+use std::iter::FromIterator;
+
+pub fn generate_ledger_from_fragments<'block0>(
+    block0_header_id: HeaderId,
+    block0_fragments: impl Iterator<Item = &'block0 Fragment>,
+    logged_fragments: impl Iterator<Item = PersistentFragmentLog>,
+) -> Result<(Ledger, VecDeque<PersistentFragmentLog>), chain_impl_mockchain::ledger::Error> {
+    let mut ledger = Ledger::new(block0_header_id, block0_fragments)?;
+    let parameters = ledger.get_ledger_parameters();
+    let block_date = ledger.date();
+    let mut fragments = VecDeque::from_iter(logged_fragments);
+    println!("{}", fragments.len());
+    let mut counter = 0;
```

Comment:
What is the idea behind using this counter? I am not sure I follow the logic here.

Reply:
Avoid looping indefinitely if there are some fragments that are not valid (e.g. a spending counter is skipped)